### ISSUE TITLE
"502 Bad Gateway nginx/1.4.7" when I use SSL

### ISSUE THREAD

**Issue by nickretallack** [inconshreveable/ngrok#123]
I created a self-signed SSL certificate and attempted to access my app through ngrok.  It gives me [this page](https://www.dropbox.com/s/2umnhd36736pepk/Screenshot%202014-04-23%2019.03.14.png).  I'm not using nginx, so it must be ngrok serving up this page.

**Comment by MemorableUsername** [comment]
Sorry to exhume this old issue from the grave, but I'm currently tackling a problem where I need to preserve an HTTPS connection all the way through the client and I've found the information in this issue very helpful and relevant.

I've set up an ngrok server of my own with the help of [this guide](http://www.svenbit.com/2014/09/run-ngrok-on-your-own-server/). I have a number of clients that connect to my server using a custom subdomain. A specific service needs to be able to request these specific clients through their full domain name; if my server is at `internal.ngrok-server.com`, this service expects to see (for example) `my-client.internal.ngrok-server.com` and `your-client.internal.ngrok-server.com`, and any other clients that indicate their existence to the server (`*.internal.ngrok-server.com`).

This works perfectly over HTTP, but a number of my server's clients are expecting to receive HTTPS traffic. To this end, I set up the clients' standard `~/.ngrok` config as follows:

```
server_addr: internal.ngrok-server.com:4443
trust_host_root_certs: false

tunnels:
  client-name:
    proto:
      http: 80
      https: 443

```
As you've excellently documented above, HTTPS traffic appears to terminate at the ngrok server itself rather than the client; thus an HTTPS request to `secretive-client.internal.ngrok-server.com` will be received as HTTP on its HTTPS port, which causes said `secretive-client` to complain.

After I read the description above, I replaced the https key for `secretive-client`'s config with the tcp key, so it looks like this:
```
...
tunnels:
  secretive-client:
    proto:
      http: 80
      tcp: 443
```

Now when `secretive-client` starts their tunnel, they server provides a port for tcp connections which they're able to use to preserve the HTTPS connection, exactly as desired!

```
ngrok                         new version available at https://ngrok.com                                    (Ctrl+C to quit)
                                                                                                                            
Tunnel Status                 online                                                                                        
Version                       1.7/1.7                                                                                       
Forwarding                    http://secretive-client.internal.ngrok-server.com -> 127.0.0.1:80                                              
Forwarding                    tcp://internal.ngrok-server.com:45899 -> 127.0.0.1:443
```

That's all well and good, and very helpful. However, due to the way the tcp domain is crafted, the specific service I mentioned earlier can no longer see this client; it expects said client to respond on `https://secretive-client.internal.ngrok-server.com`, rather than `tcp://internal.ngrok-server.com:45899`.

Is there a way I can preserve the specific subdomain the client listens on when using the tcp protocol, so the tcp address looks more like `tcp://secretive-client.internal.ngrok-server.com`? I have a foggy understanding of SSL so I understand this might not be possible, but I figure it's worth a shot to ask!
